File: Test Case Generation/gemini/cpp/test_task_10.cpp
#include <gtest/gtest.h>
#include <string>

// --- SUT (System Under Test) ---
bool is_palindrome(std::string str) {
    std::string s(str.rbegin(), str.rend());
    return s == str;
}

std::string make_palindrome(std::string str) {
    int i;
    for (i = 0; i < (int)str.length(); i++) {
        std::string rstr = str.substr(i);
        if (is_palindrome(rstr)) {
            std::string nstr = str.substr(0, i);
            std::string n2str(nstr.rbegin(), nstr.rend());
            return str + n2str;
        }
    }
    std::string n2str(str.rbegin(), str.rend());
    return str + n2str;
}

// --- Unit Tests ---

// Tests for is_palindrome function
TEST(PalindromeTest, IsPalindromeBasic) {
    EXPECT_TRUE(is_palindrome(""));
    EXPECT_TRUE(is_palindrome("a"));
    EXPECT_TRUE(is_palindrome("aba"));
    EXPECT_TRUE(is_palindrome("racecar"));
    EXPECT_FALSE(is_palindrome("abc"));
    EXPECT_FALSE(is_palindrome("at"));
}

// Tests for make_palindrome function
TEST(MakePalindromeTest, EmptyString) {
    EXPECT_EQ(make_palindrome(""), "");
}

TEST(MakePalindromeTest, SingleCharacter) {
    EXPECT_EQ(make_palindrome("a"), "a");
}

TEST(MakePalindromeTest, AlreadyAPalindrome) {
    EXPECT_EQ(make_palindrome("aba"), "aba");
    EXPECT_EQ(make_palindrome("racecar"), "racecar");
    EXPECT_EQ(make_palindrome("aa"), "aa");
}

TEST(MakePalindromeTest, ExamplesFromDocstring) {
    EXPECT_EQ(make_palindrome("cat"), "catac");
    EXPECT_EQ(make_palindrome("cata"), "catac");
}

TEST(MakePalindromeTest, GenericStrings) {
    EXPECT_EQ(make_palindrome("abc"), "abcba");
    EXPECT_EQ(make_palindrome("xyz"), "xyz yx"); // Note: substr logic
    EXPECT_EQ(make_palindrome("abcd"), "abcdcba");
}

TEST(MakePalindromeTest, PartialSuffixPalindrome) {
    // "jerry" -> suffix "y" is palindrome -> "jerr" reversed is "rrej" -> "jerryrrej"
    EXPECT_EQ(make_palindrome("jerry"), "jerryrrej");
    // "banana" -> suffix "anana" is palindrome -> "b" reversed is "b" -> "bananab"
    EXPECT_EQ(make_palindrome("banana"), "bananab");
}

TEST(MakePalindromeTest, LongestSuffixCheck) {
    // "abb" -> suffix "bb" is palindrome -> "a" reversed is "a" -> "abba"
    EXPECT_EQ(make_palindrome("abb"), "abba");
}

TEST(MakePalindromeTest, EdgeCaseNonAscii) {
    // Basic string handles bytes; ensuring no crash
    EXPECT_EQ(make_palindrome("123"), "12321");
}

File: Test Case Generation/gemini/python/test_task_10.py
import pytest

def is_palindrome(string: str) -> bool:
    """ Test if given string is a palindrome """
    return string == string[::-1]


def make_palindrome(string: str) -> str:
    """ Find the shortest palindrome that begins with a supplied string. """
    if not string:
        return ''

    beginning_of_suffix = 0

    while not is_palindrome(string[beginning_of_suffix:]):
        beginning_of_suffix += 1

    return string + string[:beginning_of_suffix][::-1]


# --- Unit Tests ---

def test_is_palindrome_empty_string():
    assert is_palindrome("") is True

def test_is_palindrome_single_char():
    assert is_palindrome("a") is True

def test_is_palindrome_valid():
    assert is_palindrome("racecar") is True
    assert is_palindrome("aba") is True

def test_is_palindrome_invalid():
    assert is_palindrome("abc") is False
    assert is_palindrome("hello") is False

def test_is_palindrome_case_sensitive():
    # Standard behavior: 'A' != 'a'
    assert is_palindrome("Abba") is False

def test_make_palindrome_empty_string():
    assert make_palindrome("") == ""

def test_make_palindrome_already_palindrome():
    assert make_palindrome("racecar") == "racecar"
    assert make_palindrome("a") == "a"
    assert make_palindrome("aa") == "aa"

def test_make_palindrome_basic():
    assert make_palindrome("cat") == "catac"
    assert make_palindrome("cata") == "catac"

def test_make_palindrome_single_char_prefix():
    # Longest suffix palindrome is 'cc', prefix is 'a'
    assert make_palindrome("acc") == "acca"

def test_make_palindrome_no_suffix_palindrome():
    # Longest suffix palindrome is 'd', prefix 'abc'
    assert make_palindrome("abcd") == "abcdcba"

def test_make_palindrome_long_string():
    assert make_palindrome("xyz") == "xyzya" or "xyz" + "yx"
    assert make_palindrome("xyz") == "xyzyx"

def test_make_palindrome_with_spaces():
    assert make_palindrome("race car") == "race car acecar"

def test_make_palindrome_numeric_strings():
    assert make_palindrome("123") == "12321"